compiler/runtime/OMRRelocationTarget.hpp

```diff
@@ -0,0 +1,87 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef OMR_RELOCATION_TARGET_INCL
+#define OMR_RELOCATION_TARGET_INCL
+
+#ifndef OMR_RELOCATION_TARGET_CONNECTOR
+#define OMR_RELOCATION_TARGET_CONNECTOR
+namespace OMR { class RelocationTarget; }
+namespace OMR { typedef OMR::RelocationTarget RelocationTargetConnector; }
+#endif
+
+
+#include <stddef.h>
+#include <stdint.h>
+#include "env/TRMemory.hpp"
+#include "env/jittypes.h"
+#include "env/JitConfig.hpp"
+
+namespace TR 
+{
+class RelocationRecord;
+class RelocationRuntime;
+class RelocationTarget;
+}
+
+namespace OMR
+{
+
+class  OMR_EXTENSIBLE RelocationTarget
+   {
+public:
+   TR_ALLOC(TR_Memory::Relocation)
+   void * operator new(size_t, TR::JitConfig *);
+   RelocationTarget(TR::RelocationRuntime *reloRuntime)
+      {
+      _reloRuntime = reloRuntime;
+      }
+   TR::RelocationTarget* self();
+   TR::RelocationRuntime *reloRuntime()                                   { return _reloRuntime; }
+   bool isOrderedPairRelocation(TR::RelocationRecord *reloRecord, TR::RelocationTarget *reloTarget);
+   void flushCache(uint8_t *codeStart, unsigned long size)       {} // default impl is empty
+   void preRelocationsAppliedEvent()                             {} // default impl is empty
+   uintptr_t loadRelocationRecordValue(uintptr_t *address)     { return *address; }
```

Comment:
I don't understand what `loadRelocationRecordValue` means for the `RelocationTarget` .. it looks like a synonym for `loadPointer` (same for `storeRelocationRecordValue` and `storePointer`).  What is a "relocation record value" ?

Reply:
It is used to distinguish between the address and relocation-specific value, when it's loaded from a Binary Template.  I think relocations are quite project specific, The example can be observed [here in openj9](https://github.com/eclipse-openj9/openj9/blob/54e151112585b88a53b463fc8df96e9418ee8fef/runtime/compiler/runtime/RelocationRecord.cpp#L1102). I can either add some explanation and this link to the code, or I can remove the method, What would you suggest?

Reply:
It seems like an abstraction that doesn't have a very clear purpose to me. At the end of the day both load/store pointer sized values and are public. The linked example in OpenJ9 could just as easily use `loadPointer` couldn't it? Is it clear to a user when to use `loadRelocationRecordValue` and when to use `loadPointer`? If so it should be documented somewhere, but if not then it should probably be removed.

Reply:
(I will push the changes shortly and request the review again after)

Reply:
OK. I think I've put it together. The generic parameters and similar implementations make it difficult to understand what the difference is. I have to look at the usage of both to discern your intentions and there's nothing preventing someone from using the wrong one and getting seemingly correct results. If possible, the parameters to each should make it difficult or impossible to use one where the other is intended.

Having said that, I understand that the code you're referring to in OpenJ9 is equally confusing and you aren't designing everything from first principles but rather adapting what's in OpenJ9 to OMR.

This code in OpenJ9, which seems to use `loadUnsigned16b()` to access data in the record, goes against what you're saying above, right? Ideally an overload of `loadRelocationRecordValue()` that can load unsigned 16-bit values should exist and be used? (I'm not suggesting you add one, just trying to understand your intentions.)

```
uint16_t
TR_RelocationRecordValidateClassByName::beholderID(TR_RelocationTarget *reloTarget)
   {
   return reloTarget->loadUnsigned16b((uint8_t *) &((TR_RelocationRecordValidateClassByNameBinaryTemplate *)_record)->_beholderID);
   }

...

uintptr_t
TR_RelocationRecordValidateClassByName::classChainOffset(TR_RelocationTarget *reloTarget)
   {
   return reloTarget->loadRelocationRecordValue((uintptr_t *) &((TR_RelocationRecordValidateClassByNameBinaryTemplate *)_record)->_classChainOffsetInSCC);
   }
```

Reply:
The facilities in the `RelocationTarget` should correspond to the kinds of values written by the platform's code generator into the code (the ones that may need to be relocated). The base relocation target facilities in OpenJ9 provide ways to write bytes, shorts, ints, pointers, etc. It gets more interesting when you look at how, for example, a method pointer may be written into the code cache. For x86, it's probably a 64-bit constant sitting in either an instruction or in a data pool. For POWER, 4 separate nibbles (shorts) are spread through a 5 instruction sequence. To the `RelocationRecord`, it calls `storeMethodPointer()` (or something like that, haven't looked at the code in a while). The platform's `RelocationTarget` knows if that's writing a 64-bit constant at the relocation offset or writing 4 separate nibbles at designated offsets to the base relocation offset. I don't really know how to interpret `loadRelocationRecordValue` because it's not even a relocation *record* value that's being read/written.

I think, for starters, all you need are the simple "C" type instance functions from OpenJ9's relocation target, which you've got below. Projects can extend with their own "artifacts" that need to be read/written to the code/data.